File: typescript/src/identity/EridDirectory.ts
/**
 * EridDirectory — resolves rotating {@link EphemeralRoutingId} wire addresses to and from the stable
 * peer identities behind them. A node shares its secret routingKey with a peer INSIDE the Signal
 * session; each side stores the other's key here, so either can compute the other's current ERID and
 * reverse-resolve an inbound ERID back to the peer. An outsider holds no key and can do neither.
 *
 * Port of the C# reference (src/AetherNet.Core/Identity/EridDirectory.cs).
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_EPOCH_SECONDS, DEFAULT_LENGTH, deriveForEpoch, epochFor } from "./EphemeralRoutingId.js";

export class EridDirectory {
  readonly #myRoutingKey: Uint8Array;
  readonly #epochSeconds: number;
  readonly #eridLength: number;
  readonly #peerKeys = new Map<string, Uint8Array>();

  /**
   * @param myRoutingKey this node's secret routingKey (from {@link deriveRoutingKey}). Copied.
   */
  constructor(
    myRoutingKey: Uint8Array,
    epochSeconds: number = DEFAULT_EPOCH_SECONDS,
    eridLength: number = DEFAULT_LENGTH,
  ) {
    if (myRoutingKey.length === 0) throw new Error("ERID: myRoutingKey cannot be empty");
    if (epochSeconds <= 0) throw new Error("ERID: epochSeconds must be positive");
    this.#myRoutingKey = myRoutingKey.slice();
    this.#epochSeconds = epochSeconds;
    this.#eridLength = eridLength;
  }

  /** Our own current ERID for the epoch containing `unixSeconds`. */
  myErid(unixSeconds: number): string {
    return deriveForEpoch(this.#myRoutingKey, epochFor(unixSeconds, this.#epochSeconds), this.#eridLength);
  }

  /** Store a peer's routingKey, learned inside an established session. Idempotent. */
  rememberPeer(peerUhid: string, peerRoutingKey: Uint8Array): void {
    if (!peerUhid) throw new Error("ERID: peerUhid cannot be empty");
    if (peerRoutingKey.length === 0) throw new Error("ERID: peerRoutingKey cannot be empty");
    this.#peerKeys.set(peerUhid, peerRoutingKey.slice());
  }

  /** Forget a peer (session torn down / excommunicated). Returns false if unknown. */
  forgetPeer(peerUhid: string): boolean {
    return this.#peerKeys.delete(peerUhid);
  }

  /** The current ERID a known peer presents this epoch, or null if we hold no key for them. */
  eridForPeer(peerUhid: string, unixSeconds: number): string | null {
    const key = this.#peerKeys.get(peerUhid);
    return key ? deriveForEpoch(key, epochFor(unixSeconds, this.#epochSeconds), this.#eridLength) : null;
  }

  /**
   * Reverse-resolve an inbound wire ERID to the stable peer UHID behind it for the given epoch, or
   * null if no known peer currently presents it. O(n) over known peers (a node's relationship count).
   */
  resolvePeer(erid: string, unixSeconds: number): string | null {
    if (!erid) return null;
    const epoch = epochFor(unixSeconds, this.#epochSeconds);
    for (const [uhid, key] of this.#peerKeys) {
      if (deriveForEpoch(key, epoch, this.#eridLength) === erid) return uhid;
    }
    return null;
  }

  /** Number of peers whose routingKey we currently hold. */
  get knownPeerCount(): number {
    return this.#peerKeys.size;
  }
}

File: rust/src/identity/erid_directory.rs
// SPDX-License-Identifier: MIT

//! Resolves rotating ERID ([`super::ephemeral_routing_id`]) wire addresses to and from
//! the stable peer identities behind them — the piece that lets an ESTABLISHED
//! relationship follow a peer's rotating address while an outsider cannot.
//!
//! A node derives its OWN secret routing key once (via [`super::derive_routing_key`]) and
//! shares it with a peer INSIDE the established Signal session — never on the wire. Each
//! side stores the other's routing key here, so either can compute the other's current
//! ERID for addressing, and reverse-resolve an inbound ERID back to the peer it belongs
//! to. An outsider holds no routing key and can do neither. Port of the C# reference
//! (`src/AetherNet.Core/Identity/EridDirectory.cs`).

use std::collections::HashMap;

use super::ephemeral_routing_id::{
    derive, EphemeralRoutingIdError, DEFAULT_EPOCH_SECONDS, DEFAULT_LENGTH,
};

/// In-memory directory mapping peer UHIDs to their secret routing keys.
#[derive(Debug, Clone)]
pub struct EridDirectory {
    my_routing_key: Vec<u8>,
    epoch_seconds: i64,
    erid_length: usize,
    peer_keys: HashMap<String, Vec<u8>>,
}

impl EridDirectory {
    /// Create a directory for a node holding `my_routing_key`, using the default rotation
    /// window and ERID length.
    ///
    /// # Panics
    /// Panics if `my_routing_key` is empty — a programming error; derive it first via
    /// [`super::derive_routing_key`].
    #[must_use]
    pub fn new(my_routing_key: &[u8]) -> Self {
        Self::with_params(my_routing_key, DEFAULT_EPOCH_SECONDS, DEFAULT_LENGTH)
    }

    /// Create a directory with explicit rotation parameters.
    ///
    /// # Panics
    /// Panics if `my_routing_key` is empty or `epoch_seconds <= 0`.
    #[must_use]
    pub fn with_params(my_routing_key: &[u8], epoch_seconds: i64, erid_length: usize) -> Self {
        assert!(!my_routing_key.is_empty(), "my_routing_key cannot be empty");
        assert!(epoch_seconds > 0, "epoch_seconds must be positive");
        Self {
            my_routing_key: my_routing_key.to_vec(),
            epoch_seconds,
            erid_length,
            peer_keys: HashMap::new(),
        }
    }

    /// Our own current ERID for the epoch containing `unix_seconds` — the address we
    /// present on the wire this window.
    ///
    /// # Errors
    /// Propagates [`EphemeralRoutingIdError`] from the underlying derivation.
    pub fn my_erid(&self, unix_seconds: i64) -> Result<String, EphemeralRoutingIdError> {
        derive(&self.my_routing_key, unix_seconds, self.epoch_seconds, self.erid_length)
    }

    /// Store a peer's routing key, learned inside an established session. Idempotent; a
    /// later call replaces an earlier key for the same peer.
    ///
    /// # Panics
    /// Panics if `peer_uhid` or `peer_routing_key` is empty.
    pub fn remember_peer(&mut self, peer_uhid: &str, peer_routing_key: &[u8]) {
        assert!(!peer_uhid.is_empty(), "peer_uhid cannot be empty");
        assert!(!peer_routing_key.is_empty(), "peer_routing_key cannot be empty");
        self.peer_keys
            .insert(peer_uhid.to_string(), peer_routing_key.to_vec());
    }

    /// Forget a peer (session torn down / excommunicated). Returns `false` if unknown.
    pub fn forget_peer(&mut self, peer_uhid: &str) -> bool {
        self.peer_keys.remove(peer_uhid).is_some()
    }

    /// The current ERID a known peer presents this epoch, or `None` if we hold no key
    /// for them.
    ///
    /// # Errors
    /// Propagates [`EphemeralRoutingIdError`] from the underlying derivation.
    pub fn erid_for_peer(
        &self,
        peer_uhid: &str,
        unix_seconds: i64,
    ) -> Result<Option<String>, EphemeralRoutingIdError> {
        match self.peer_keys.get(peer_uhid) {
            Some(key) => Ok(Some(derive(
                key,
                unix_seconds,
                self.epoch_seconds,
                self.erid_length,
            )?)),
            None => Ok(None),
        }
    }

    /// Reverse-resolve an inbound wire ERID to the stable peer UHID behind it for the
    /// given epoch, or `None` if no known peer currently presents it. O(n) over known
    /// peers — a node's actual relationship count.
    ///
    /// # Errors
    /// Propagates [`EphemeralRoutingIdError`] from the underlying derivation.
    pub fn resolve_peer(
        &self,
        erid: &str,
        unix_seconds: i64,
    ) -> Result<Option<String>, EphemeralRoutingIdError> {
        if erid.is_empty() {
            return Ok(None);
        }
        for (uhid, key) in &self.peer_keys {
            let candidate = derive(key, unix_seconds, self.epoch_seconds, self.erid_length)?;
            if candidate == erid {
                return Ok(Some(uhid.clone()));
            }
        }
        Ok(None)
    }

    /// Number of peers whose routing key we currently hold.
    #[must_use]
    pub fn known_peer_count(&self) -> usize {
        self.peer_keys.len()
    }
}
